diff --git a/code/test/Makefile b/code/test/Makefile
index 2d0e84c..6a42235 100644
--- a/code/test/Makefile
+++ b/code/test/Makefile
@@ -112,7 +112,7 @@ ifeq ($(hosttype),unknown)
 PROGRAMS = unknownhost
 else
 # change this if you create a new test program!
-PROGRAMS = add halt shell matmult sort segments test_syscall num_io char_io random str_io ascii bubblesort help create_file open_file readwrite exec test_sem sinhvien voinuoc main
+PROGRAMS = abs add halt shell matmult sort segments test_syscall num_io char_io random str_io ascii bubblesort help create_file open_file readwrite exec test_sem sinhvien voinuoc main
 endif
 
 all: $(PROGRAMS)
@@ -133,6 +133,13 @@ add: add.o start.o
 	$(LD) $(LDFLAGS) start.o add.o -o add.coff
 	$(COFF2NOFF) add.coff add
 
+abs.o: abs.c
+	$(CC) $(CFLAGS) -c abs.c
+
+abs: abs.o start.o
+	$(LD) $(LDFLAGS) start.o abs.o -o abs.coff
+	$(COFF2NOFF) abs.coff abs
+
 shell.o: shell.c
 	$(CC) $(CFLAGS) -c shell.c
 shell: shell.o start.o
diff --git a/code/test/abs.c b/code/test/abs.c
index ce75967..7ab0095 100644
--- a/code/test/abs.c
+++ b/code/test/abs.c
@@ -9,7 +9,7 @@
 
 int main() {
     int y;
-    int x=-100;
+    int x=-101;
     y = Abs(x);
     PrintNum(y);
 
diff --git a/code/userprog/exception.cc b/code/userprog/exception.cc
index a5d02c7..9f01f93 100644
--- a/code/userprog/exception.cc
+++ b/code/userprog/exception.cc
@@ -154,6 +154,21 @@ void handle_SC_Add() {
     return move_program_counter();
 }
 
+void handle_SC_Abs() {
+    DEBUG(dbgSys, "Abs " << kernel->machine->ReadRegister(4) <<  "\n");
+
+    /* Process SysAdd Systemcall*/
+    int result;
+    result = SysAbs(
+        /* int op1 */ (int)kernel->machine->ReadRegister(4));
+
+    DEBUG(dbgSys, "Add returning with " << result << "\n");
+    /* Prepare Result */
+    kernel->machine->WriteRegister(2, (int)result);
+
+    return move_program_counter();
+}
+
 void handle_SC_ReadNum() {
     int result = SysReadNum();
     kernel->machine->WriteRegister(2, result);
@@ -421,6 +436,8 @@ void ExceptionHandler(ExceptionType which) {
                     return handle_SC_Halt();
                 case SC_Add:
                     return handle_SC_Add();
+                case SC_Abs:
+                    return handle_SC_Abs();
                 case SC_ReadNum:
                     return handle_SC_ReadNum();
                 case SC_PrintNum:
diff --git a/code/userprog/ksyscall.h b/code/userprog/ksyscall.h
index 3a9a016..ebd30b1 100644
--- a/code/userprog/ksyscall.h
+++ b/code/userprog/ksyscall.h
@@ -19,7 +19,7 @@
 void SysHalt() { kernel->interrupt->Halt(); }
 
 int SysAdd(int op1, int op2) { return op1 + op2; }
-
+int SysAbs(int op1) {if(op1>0) return op1; else return -op1;}
 int SysReadNum() {
     readUntilBlank();
 
diff --git a/code/userprog/syscall.h b/code/userprog/syscall.h
index 0865b05..3e9af4e 100644
--- a/code/userprog/syscall.h
+++ b/code/userprog/syscall.h
@@ -48,7 +48,7 @@
 #define SC_Wait 52
 #define SC_Signal 53
 #define SC_GetPid 54
-
+#define SC_Abs 55
 #ifndef IN_ASM
 
 /* The system call interface.  These are the operations the Nachos
@@ -69,7 +69,7 @@ void Halt();
  */
 
 int Add(int op1, int op2);
-
+int Abs(int op1);
 int ReadNum();
 
 void PrintNum(int num);
