diff --git a/code/test/add.c b/code/test/add.c
index 70b6925..e3d20e8 100644
--- a/code/test/add.c
+++ b/code/test/add.c
@@ -9,8 +9,13 @@
 
 int main() {
     int result;
+    int i,j;
 
-    result = Add(42, 23);
+    for(i=0;i<1000;i++){
+        for(j=0;j<200000;j++);
+            PrintString("In child\n");
+        
+    }
 
     Halt();
     /* not reached */
diff --git a/code/test/exec.c b/code/test/exec.c
index feadac1..869a71d 100644
--- a/code/test/exec.c
+++ b/code/test/exec.c
@@ -8,10 +8,19 @@
 
 int main() {
     int pid;
-    pid = Exec("num_io");
+    int i,j;
+    PrintString("In parent\n");
+    pid = Exec("../test/add");
     if (pid < 0) {
         Write("Exec failed: ", 14, stdout);
         PrintNum(pid);
-    } else
+    } else{
+         for(i=0;i<1000;i++)
+         {
+                 for(j=0;j<200000;j++);
+                 PrintString("In Parent\n");
+         }
+    }
+
         Join(pid);
 }
